onnxruntime/core/providers/op_kernel_type_control.h

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <tuple>
+
+#include "boost/mp11.hpp"
+
+#include "core/common/type_list.h"
+
+/**
+ * These utilities provide a way to control what types are enabled for an Op kernel implementation.
+ *
+ * At a high level, we have the notion of supported, allowed, and enabled types.
+ * - Supported types are the types that the Op kernel implementation supports by default.
+ * - Allowed types are the types for which support is requested (for example, by external configuration).
+ * - Enabled types are the types that are supported in the actual, compiled implementation. They are obtained from the
+ *   intersection of supported and allowed types.
+ *
+ * The types are associated with an Op kernel argument. It is also possible to specify a global list of allowed types.
+ *
+ * Use of these utilities is optional. They are useful for cases where one registered Op kernel handles multiple types.
+ *
+ * See the macros below for usage details.
+ */
+
+namespace onnxruntime {
+namespace op_kernel_type_control {
+
+enum class OpKernelArgDirection {
+  Input,
+  Output
+};
+
+using OpKernelArgIndex = size_t;
+
+namespace tags {
+
+// a tag that identifies the target (Op kernel argument) of the specified types
+template <typename OpKernelTag, OpKernelArgDirection ArgDirection, OpKernelArgIndex ArgIndex>
+struct OpKernelArg {};
+
+// a tag that indicates the supported types for a particular Op kernel argument, identified by OpKernelArgTag
+template <typename OpKernelArgTag>
+struct Supported {};
+
+// a tag that indicates the allowed types for a particular Op kernel argument, identified by OpKernelArgTag
+template <typename OpKernelArgTag>
+struct Allowed {};
+
+// a tag that indicates the globally allowed types
+struct GlobalAllowed {};
+
+}  // namespace tags
+
+// holds specified Op kernel argument types
+// if types are defined, the data member 'value' should contain them in a type list
+// see the tags in onnxruntime::op_kernel_type_control::tags for intended uses
+template <typename Tag>
+struct OpKernelArgTypes {};
+
+// gives access to the enabled Op kernel argument types via the 'type' data member
+template <typename OpKernelArgTag>
+struct EnabledOpKernelArgTypes {
+ private:
+  template <typename T>
+  using GetTypeMember = typename T::type;
+
```

Comment:
would it be clearer to call it 'types'/HasTypes instead of 'type'/HasTypeMember given it's a list?

Reply:
hm, a type list is also a type. :) i don't disagree and originally had 'types', but 'type' seems more conventional, at least with examples i've seen from std

---
In reply to: [564874953](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564874953) [](ancestors = 564874953)

Reply:
I'd slightly prefer the readability of 'types' vs. having to know that 'type' is a type list given that's somewhat opaque. I would expect it's easier to comprehend what is going on with that naming if you're unfamiliar with the setup.

Are the examples in std dealing with a single type (e.g. type in a vector) or a type list?

Reply:
ok. i haven't come across examples in std dealing with type lists.

---
In reply to: [564951914](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564951914) [](ancestors = 564951914)